### ISSUE TITLE
java.lang.ClassCastException: org.apache.spark.sql.catalyst.expressions.GenericRow cannot be cast to org.apache.avro.generic.GenericRecord

### ISSUE THREAD

**Issue by heng2018** [AbsaOSS/ABRiS#56]
I tried ConfluentKafkaAvroReader using my schema in schema registration and go the following error

Caused by: java.lang.ClassCastException: org.apache.spark.sql.catalyst.expressions.GenericRow cannot be cast to org.apache.avro.generic.GenericRecord
	at za.co.absa.abris.avro.sql.AvroDeserializer$$anonfun$za$co$absa$abris$avro$sql$AvroDeserializer$$newWriter$18.apply(AvroDeserializer.scala:172)
	at za.co.absa.abris.avro.sql.AvroDeserializer$$anonfun$za$co$absa$abris$avro$sql$AvroDeserializer$$newWriter$18.apply(AvroDeserializer.scala:170)
	at za.co.absa.abris.avro.sql.AvroDeserializer$$anonfun$8.apply(AvroDeserializer.scala:332)
	at za.co.absa.abris.avro.sql.AvroDeserializer$$anonfun$8.apply(AvroDeserializer.scala:328)
	at za.co.absa.abris.avro.sql.AvroDeserializer$$anonfun$getRecordWriter$1.apply(AvroDeserializer.scala:350)
	at za.co.absa.abris.avro.sql.AvroDeserializer$$anonfun$getRecordWriter$1.apply(AvroDeserializer.scala:347)
	at za.co.absa.abris.avro.sql.AvroDeserializer$$anonfun$3.apply(AvroDeserializer.scala:60)
	at za.co.absa.abris.avro.sql.AvroDeserializer$$anonfun$3.apply(AvroDeserializer.scala:58)
	at za.co.absa.abris.avro.sql.AvroDeserializer.deserialize(AvroDeserializer.scala:74)
	at za.co.absa.abris.avro.sql.AvroDataToCatalyst.nullSafeEval(AvroDataToCatalyst.scala:59)
	... 17 more

**Comment by Inylschek** [comment]
I got the exact same error with 2_11, 3.0.1 when calling from_confluent_avro with a schema string.

Not sure if it's a bug or if I'm not using this correctly.

**Comment by cerveada** [comment]
Hi, thanks for repporting the error. Could any of you please provide more information? What schema were you using? How have you serialized the data?

**Comment by Inylschek** [comment]
My schema is very complex. Not sure about serialization. Just reading from a Kafka topic. 

I'm just investigating what happens for a simpler schema first before coming back with more details...

**Comment by Inylschek** [comment]
Not sure how the data is serialized, however I have managed to read from a different Kafka topic with a different avro schema. This schema is much simpler than the schema that fails.

I amn't sure which serialization is being used, but I know that both the simple and complex schema are written in the same way. 

Here is the complex schema that is failing

```
root
 |-- from_avro(value): struct (nullable = true)
 |    |-- a: string (nullable = false)
 |    |-- b: string (nullable = false)
 |    |-- c: string (nullable = false)
 |    |-- d: long (nullable = false)
 |    |-- e: integer (nullable = false)
 |    |-- f: struct (nullable = false)
 |    |    |-- g: struct (nullable = false)
 |    |    |    |-- value: double (nullable = true)
 |    |    |    |-- tsMs: long (nullable = false)
 |    |    |    |-- status: integer (nullable = false)
 |    |    |    |-- changed: boolean (nullable = false)
 |    |    |-- h: struct (nullable = false)
 |    |    |    |-- value: boolean (nullable = true)
 |    |    |    |-- tsMs: long (nullable = false)
 |    |    |    |-- status: integer (nullable = false)
 |    |    |    |-- changed: boolean (nullable = false)
 |    |    |-- i: struct (nullable = false)
 |    |    |    |-- value: string (nullable = true)
 |    |    |    |-- tsMs: long (nullable = false)
 |    |    |    |-- status: integer (nullable = false)
 |    |    |    |-- changed: boolean (nullable = false)
 |    |    |-- j: struct (nullable = false)
 |    |    |    |-- value: boolean (nullable = true)
 |    |    |    |-- tsMs: long (nullable = false)
 |    |    |    |-- status: integer (nullable = false)
 |    |    |    |-- changed: boolean (nullable = false)
 |    |    |-- k: struct (nullable = false)
 |    |    |    |-- value: integer (nullable = false)
 |    |    |    |-- tsMs: long (nullable = false)
 |    |    |    |-- status: integer (nullable = false)
 |    |    |    |-- changed: boolean (nullable = false)
 |    |    |-- l: struct (nullable = false)
 |    |    |    |-- value: boolean (nullable = true)
 |    |    |    |-- tsMs: long (nullable = false)
 |    |    |    |-- status: integer (nullable = false)
 |    |    |    |-- changed: boolean (nullable = false)
 |    |    |-- m: struct (nullable = false)
 |    |    |    |-- value: boolean (nullable = true)
 |    |    |    |-- tsMs: long (nullable = false)
 |    |    |    |-- status: integer (nullable = false)
 |    |    |    |-- changed: boolean (nullable = false)
 |    |    |-- n: struct (nullable = false)
 |    |    |    |-- value: boolean (nullable = true)
 |    |    |    |-- tsMs: long (nullable = false)
 |    |    |    |-- status: integer (nullable = false)
 |    |    |    |-- changed: boolean (nullable = false)
 |    |    |-- o: struct (nullable = false)
 |    |    |    |-- value: boolean (nullable = true)
 |    |    |    |-- tsMs: long (nullable = false)
 |    |    |    |-- status: integer (nullable = false)
 |    |    |    |-- changed: boolean (nullable = false)
 |    |    |-- p: struct (nullable = false)
 |    |    |    |-- value: boolean (nullable = true)
 |    |    |    |-- tsMs: long (nullable = false)
 |    |    |    |-- status: integer (nullable = false)
 |    |    |    |-- changed: boolean (nullable = false)
 |    |    |-- q: struct (nullable = false)
 |    |    |    |-- value: integer (nullable = true)
 |    |    |    |-- tsMs: long (nullable = false)
 |    |    |    |-- status: integer (nullable = false)
 |    |    |    |-- changed: boolean (nullable = false)
 |    |    |-- r: struct (nullable = false)
 |    |    |    |-- value: integer (nullable = true)
 |    |    |    |-- tsMs: long (nullable = false)
 |    |    |    |-- status: integer (nullable = false)
 |    |    |    |-- changed: boolean (nullable = false)
 |    |    |-- s: struct (nullable = false)
 |    |    |    |-- value: float (nullable = true)
 |    |    |    |-- tsMs: long (nullable = false)
 |    |    |    |-- status: integer (nullable = false)
 |    |    |    |-- changed: boolean (nullable = false)
 |    |    |-- t: struct (nullable = false)
 |    |    |    |-- value: double (nullable = true)
 |    |    |    |-- tsMs: long (nullable = false)
 |    |    |    |-- status: integer (nullable = false)
 |    |    |    |-- changed: boolean (nullable = false)
 |    |    |-- u: struct (nullable = false)
 |    |    |    |-- value: integer (nullable = true)
 |    |    |    |-- tsMs: long (nullable = false)
 |    |    |    |-- status: integer (nullable = false)
 |    |    |    |-- changed: boolean (nullable = false)

```

I can confirm that the kafka topic containing the complex schema is being read by a microservice without any problems, so the data itself is not corrupt.

**Comment by Inylschek** [comment]
Also, we are using Confluent's schema registry for encoding, however I am still using from_confluent_avro with a jsonSchema string derived by reading the avsc file. In other words, the jsonSchema I am using does not include any additional information that Confluent adds

**Comment by heng2018** [comment]
we use confluent schema registration
our code originally run on databricks notebook and we use databricks avro.functions.from_avro. 
now we migrate the code off databricks and replace that part with ABRiS from_confluent_avro with schema registry signature, same producer , topic and messages and got this error

**Comment by Inylschek** [comment]
I've just seen on the wiki that abris does not support batch processing which is what I am trying to do (but I did get it to work for the simple schema!)

Interested to know if that could be the problem? Even if it is, heng2018 still has a problem here.

**Comment by cerveada** [comment]
> I've just seen on the wiki that abris does not support batch processing which is what I am trying to do (but I did get it to work for the simple schema!)

The ```WIKI.wiki``` file is outdated. It is about the old API from Abris 2.x.x. The current API should work anywhere where you can do select. 

**Good news** is that I was able to simulate this exception. The fix will be part of next release.

**Comment by heng2018** [comment]
great news!
when will the the next release?

**Comment by cerveada** [comment]
Most probably this week.

**Comment by cerveada** [comment]
Fixed in 3.0.2
